File: sdks/cpp/connections/REST/examples/one_of_everything_REST/one_of_everything_REST.cpp
/*
 * Copyright 2025 Ross Video Ltd
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS”
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * RE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * @brief Example program containing one of everyhting.
 * @file one_of_everything.cpp
 * @copyright Copyright © 2025 Ross Video Ltd
 * @author Benjamin.whitten@rossvideo.com
 */

// device model
#include "device.one_of_everything.yaml.h" 

//common
#include <utils.h>
#include <Device.h>
#include <ParamWithValue.h>
#include <ParamDescriptor.h>

// REST
#include <ServiceImpl.h>

#include "absl/flags/parse.h"
#include "absl/flags/usage.h"
#include "absl/strings/str_format.h"

#include <iomanip>
#include <iostream>
#include <memory>
#include <regex>
#include <stdexcept>
#include <string>
#include <thread>
#include <chrono>
#include <signal.h>
#include <functional>
#include "Logger.h"


using catena::REST::CatenaServiceImpl;

using namespace catena::common;


CatenaServiceImpl *globalApi = nullptr;
std::atomic<bool> fibLoop = false;
std::unique_ptr<std::thread> fibThread = nullptr;
std::atomic<bool> counterLoop = true;

// handle SIGINT
void handle_signal(int sig) {
    std::thread t([sig]() {
        DEBUG_LOG << "Caught signal " << sig << ", shutting down";
        fibLoop = false;
        counterLoop = false;
        if (globalApi != nullptr) {
            globalApi->Shutdown();
            globalApi = nullptr;
        }
    });
    t.join();
}

void defineCommands() {
    catena::exception_with_status err{"", catena::StatusCode::OK};

    // Use an oid to get a pointer to the command you want to define
    // In Catena, commands have IParam type
    std::unique_ptr<IParam> fibStart = dm.getCommand("/fib_start", err);
    assert(fibStart != nullptr);

    /* 
     * Define a lambda function to be executed when the command is called
     * The lambda function must take a catena::Value as an argument and return 
     * a catena::CommandResponse.
     * Starts a thread which updates the number_example parameter with the next
     * number of the Fibonacci sequence every second.
     */
    fibStart->defineCommand([](catena::Value value) -> std::unique_ptr<IParamDescriptor::ICommandResponder> { 
        return std::make_unique<ParamDescriptor::CommandResponder>([value]() -> ParamDescriptor::CommandResponder {
            catena::exception_with_status err{"", catena::StatusCode::OK};
            catena::CommandResponse response;
            std::unique_ptr<IParam> intParam = dm.getParam("/number_example", err);
            
            // If the loop is already running, return an exception.
            if (fibThread) {
                response.mutable_exception()->set_type("Invalid Command");
                response.mutable_exception()->set_details("Already running");
            // If the state parameter does not exist, return an exception.
            } else if (!intParam) {
                response.mutable_exception()->set_type("Invalid Command");
                response.mutable_exception()->set_details(err.what());
            } else {
                fibLoop = true;
                // Detaching thread to update number_example with next number of
                // the Fibonacci sequence every second.
                fibThread = std::make_unique<std::thread>([intParam = std::move(intParam)]() {
                    auto& fibParam = *dynamic_cast<ParamWithValue<int32_t>*>(intParam.get());
                    uint32_t prev = 0;
                    uint32_t curr = 1;
                    while (fibLoop) {
                        std::this_thread::sleep_for(std::chrono::seconds(1));
                        uint32_t next = prev + curr;
                        prev = curr;
                        curr = next;
                        {
                        std::lock_guard lg(dm.mutex());
                        fibParam.get() = next;
                        dm.valueSetByServer.emit("/number_example", &fibParam);
                        }
                    }
                });

                DEBUG_LOG << "Fibonacci sequence start";
                response.mutable_no_response();
            }
            co_return response;
        }());
    });

    // This stops the looping thread in the fib_start command above.
    std::unique_ptr<IParam> fibStop = dm.getCommand("/fib_stop", err);
    assert(fibStop != nullptr);
    fibStop->defineCommand([](catena::Value value) -> std::unique_ptr<IParamDescriptor::ICommandResponder> { 
        return std::make_unique<ParamDescriptor::CommandResponder>([value]() -> ParamDescriptor::CommandResponder {
            catena::exception_with_status err{"", catena::StatusCode::OK};
            catena::CommandResponse response;
            if (fibThread) {
                fibLoop = false;
                fibThread->join();
                fibThread = nullptr;
                DEBUG_LOG << "Fibonacci sequence stop";
                response.mutable_no_response();
            } else {
                response.mutable_exception()->set_type("Invalid Command");
                response.mutable_exception()->set_details("Already stopped");
            }
            co_return response;
        }());
    });

    // This sets the value of number_example.
    std::unique_ptr<IParam> fibSet = dm.getCommand("/fib_set", err);
    assert(fibSet != nullptr);
    fibSet->defineCommand([](catena::Value value) -> std::unique_ptr<IParamDescriptor::ICommandResponder> { 
        return std::make_unique<ParamDescriptor::CommandResponder>([value]() -> ParamDescriptor::CommandResponder {
            catena::exception_with_status err{"", catena::StatusCode::OK};
            catena::CommandResponse response;
            std::unique_ptr<IParam> intParam = dm.getParam("/number_example", err);
            if (intParam) {
                auto& fibParam = *dynamic_cast<ParamWithValue<int32_t>*>(intParam.get());
                fibParam.get() = value.int32_value();
                dm.valueSetByServer.emit("/number_example", &fibParam);
                response.mutable_no_response();
            } else {
                response.mutable_exception()->set_type("Invalid Command");
                response.mutable_exception()->set_details("/number_example not found");
            }
            co_return response;
        }());
    });

    // This fills float_array with 128 random floats rounded to 3 decimal places.
    std::unique_ptr<IParam> randomize = dm.getCommand("/randomize", err);
    assert(randomize != nullptr);
    randomize->defineCommand([](catena::Value value) -> std::unique_ptr<IParamDescriptor::ICommandResponder> { 
        return std::make_unique<ParamDescriptor::CommandResponder>([value]() -> ParamDescriptor::CommandResponder {
            catena::exception_with_status err{"", catena::StatusCode::OK};
            catena::CommandResponse response;
            std::unique_ptr<IParam> floatArray = dm.getParam("/float_array", err);
            if (!floatArray) {
                response.mutable_exception()->set_type("Invalid Command");
                response.mutable_exception()->set_details(err.what());
            } else {
                auto& floatArrayR = *dynamic_cast<ParamWithValue<std::vector<float>>*>(floatArray.get());
                std::vector<float>& randomArray = floatArrayR.get();
                randomArray.clear();
                // Generates random floats between 0 and 80 and rounds to 3 decimal places.
                for (int i = 0; i < floatArrayR.getDescriptor().max_length(); i++) {
                    float randomNum = static_cast<float>(rand()) / static_cast<float>(RAND_MAX / 80);
                    randomNum = std::round(randomNum * 1000) / 1000;
                    randomArray.push_back(randomNum);
                }
                DEBUG_LOG << "Randomized float array";
                response.mutable_no_response();
            }
            
            co_return response;
        }());
    });

    // This simulates a tape bot and returns a stream of responses.
    std::unique_ptr<IParam> tapeBot = dm.getCommand("/tape_bot", err);
    assert(tapeBot != nullptr);
    tapeBot->defineCommand([](catena::Value value) -> std::unique_ptr<IParamDescriptor::ICommandResponder> { 
        return std::make_unique<ParamDescriptor::CommandResponder>([value]() -> ParamDescriptor::CommandResponder {
            catena::exception_with_status err{"", catena::StatusCode::OK};
            catena::CommandResponse response;
            // Locating
            DEBUG_LOG << "Locating tape...";
            response.mutable_response()->set_string_value("Locating tape...");
            co_yield response;
            std::this_thread::sleep_for(std::chrono::seconds(2));
            // Loading
            DEBUG_LOG << "Tape found, loading...";
            response.Clear();
            response.mutable_response()->set_string_value("Tape found, loading...");
            co_yield response;
            std::this_thread::sleep_for(std::chrono::seconds(2));
            // Seeking
            DEBUG_LOG << "Tape loaded, seeking...";
            response.Clear();
            response.mutable_response()->set_string_value("Tape loaded, seeking...");
            co_yield response;
            std::this_thread::sleep_for(std::chrono::seconds(2));
            // Reading
            DEBUG_LOG << "File loaded, reading...";
            response.Clear();
            response.mutable_response()->set_string_value("File loaded, reading...");
            co_yield response;
            std::this_thread::sleep_for(std::chrono::seconds(2));
            // Done
            DEBUG_LOG << "File loaded.";
            response.Clear();
            response.mutable_response()->set_string_value("File loaded.");
            co_return response;
        }());
    });
}

// Starts a loop on a detached thread that updates the counter parameter by 1
// every second. Resets to 0 when the counter reaches 200.
void startCounter() {
    catena::exception_with_status err{"", catena::StatusCode::OK};
    // The rest is the "sending end" of the status update example
    std::unique_ptr<IParam> param = dm.getParam("/counter", err);
    if (param == nullptr) {
        throw err;
    }
    // downcast the IParam to a ParamWithValue<int32_t>
    auto& counter = *dynamic_cast<ParamWithValue<int32_t>*>(param.get());
    counter.get() = 0; // Initialize counter to 0
    while (counterLoop) {
        // update the counter once per second, and emit the event
        std::this_thread::sleep_for(std::chrono::seconds(1));
        {
            std::lock_guard lg(dm.mutex());
            if (counter.get()++ >= 200) {
                counter.get() = 0; // Reset counter to 0 when it reaches 200
            }
            DEBUG_LOG << counter.getOid() << " set to " << counter.get();
            dm.valueSetByServer.emit("/counter", &counter);
        }
    }
}

void RunRESTServer() {
    // install signal handlers
    signal(SIGINT, handle_signal);
    signal(SIGTERM, handle_signal);
    signal(SIGKILL, handle_signal);

    try {
        // Getting flags.
        std::string EOPath = absl::GetFlag(FLAGS_static_root);
        bool authorization = absl::GetFlag(FLAGS_authz);
        uint16_t port = absl::GetFlag(FLAGS_port);

        // Creating and running the REST service.
        CatenaServiceImpl api({&dm}, EOPath, authorization, port);
        globalApi = &api;
        DEBUG_LOG << "API Version: " << api.version();
        DEBUG_LOG << "REST on 0.0.0.0:" << port;

        std::thread loop(startCounter);

        api.run();

        if (fibThread) { fibThread->join(); }
        loop.join();
    } catch (std::exception &why) {
        LOG(ERROR) << "Problem: " << why.what();
    }
}

int main(int argc, char* argv[])
{
    FLAGS_logtostderr = false;          // Keep logging to files
    FLAGS_log_dir = GLOG_LOGGING_DIR;   // Set the log directory
    google::InitGoogleLogging("one_of_everything_REST");

    std::string addr;
    absl::SetProgramUsageMessage("Runs the Catena Service");
    absl::ParseCommandLine(argc, argv);
  
    addr = absl::StrFormat("0.0.0.0:%d", absl::GetFlag(FLAGS_port));
  
    // commands should be defined before starting the RPC server 
    defineCommands();

    std::thread catenaRestThread(RunRESTServer);
    catenaRestThread.join();
    return 0;
}

File: sdks/go/examples/one_of_everything_REST/one_of_everything_REST.go
/*
 * Copyright 2026 Ross Video Ltd
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * RE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * @brief Example program containing one of everything.
 * @file one_of_everything_REST.go
 * @copyright Copyright © 2026 Ross Video Ltd
 * @author Nelson Daniels (nelson.daniels@rossvideo.com)
 * @date 2026-01-13
 */

package main

import (
	"bytes"
	"embed"
	"fmt"
	"io"
	"io/fs"
	"mime"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/rossvideo/catena/sdks/go/pkg/catena"
	"github.com/rossvideo/catena/sdks/go/pkg/logger"
	"github.com/rossvideo/catena/sdks/go/pkg/rest"
)

//go:embed static/*
var staticFS embed.FS

// Asset represents a binary asset with metadata
type Asset struct {
	ContentType string
	Data        []byte
	Filename    string
}

// CommandHandler processes a command and returns a response
type CommandHandler func(payload any) catena.StatusResult

// Counter state with thread-safe operations
type CounterState struct {
	mu      sync.RWMutex
	value   int64
	running bool
}

func (c *CounterState) GetValue() int64 {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.value
}

func (c *CounterState) IsRunning() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.running
}

func (c *CounterState) Start() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.running = true
}

func (c *CounterState) Stop() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.running = false
}

func (c *CounterState) Add(n int64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value += n
}

func (c *CounterState) Reset() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value = 0
}

func (c *CounterState) Increment() {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.running {
		c.value++
	}
}

// Global state for graceful shutdown
var (
	shutdownChan = make(chan struct{})
	srv          *rest.Server
)

func main() {
	cfg := logger.ParseConfigWithVerbosity("CATENA")
	cfg.AppName = "one_of_everything_REST"

	if err := logger.Init(cfg); err != nil {
		fmt.Fprintf(os.Stderr, "failed to initialize logger: %v\n", err)
		os.Exit(1)
	}
	defer logger.Close()

	// Handle signals for graceful shutdown
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		sig := <-sigChan
		logger.Info("Caught signal, shutting down", "signal", sig)
		close(shutdownChan)
	}()

	portStr := envOr("CATENA_PORT", "6254")
	port, err := strconv.Atoi(portStr)
	if err != nil {
		logger.Error("invalid CATENA_PORT", "error", err)
		os.Exit(1)
	}

	// ==========================================================================
	// Counter State
	// ==========================================================================
	counter := &CounterState{}

	// Start counter goroutine - increments every second when running
	go func() {
		ticker := time.NewTicker(1 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				if counter.IsRunning() {
					counter.Increment()
					logger.Info("Counter tick", "value", counter.GetValue())
				}
			case <-shutdownChan:
				return
			}
		}
	}()

	// ==========================================================================
	// Device Metadata (for GetDevice endpoint)
	// ==========================================================================
	devices := map[int]map[string]any{
		0: {
			"slot":              0,
			"multi_set_enabled": true,
			"detail_level":      "FULL",
			"access_scopes":     []string{"monitor", "operate", "configure", "administer"},
			"default_scope":     "operate",
		},
	}

	// ==========================================================================
	// Parameters (for GetParam/SetParam endpoints)
	// ==========================================================================
	params := &sync.Map{}

	// Video parameters
	params.Store("video/resolution", "1920x1080")
	params.Store("video/brightness", 50)
	params.Store("video/contrast", 50)
	params.Store("video/saturation", 50)

	// Audio parameters
	params.Store("audio/volume", 75)
	params.Store("audio/muted", false)

	// System parameters
	params.Store("system/device_name", "Demo Device")

	// ==========================================================================
	// Commands (for ExecuteCommand endpoint)
	// ==========================================================================

	// Helper to build counter response
	buildCounterResponse := func() map[string]any {
		return map[string]any{
			"counter": counter.GetValue(),
			"running": counter.IsRunning(),
		}
	}

	commands := map[string]CommandHandler{
		// Counter commands
		"start": func(payload any) catena.StatusResult {
			if counter.IsRunning() {
				logger.Info("Start command - already running")
				return catena.OK(buildCounterResponse())
			}
			counter.Start()
			logger.Info("Counter started", "value", counter.GetValue())
			return catena.OK(buildCounterResponse())
		},

		"stop": func(payload any) catena.StatusResult {
			if !counter.IsRunning() {
				logger.Info("Stop command - already stopped")
				return catena.OK(buildCounterResponse())
			}
			counter.Stop()
			logger.Info("Counter stopped", "value", counter.GetValue())
			return catena.OK(buildCounterResponse())
		},

		"add10": func(payload any) catena.StatusResult {
			counter.Add(10)
			logger.Info("Added 10 to counter", "value", counter.GetValue())
			return catena.OK(buildCounterResponse())
		},

		"reset": func(payload any) catena.StatusResult {
			counter.Reset()
			logger.Info("Counter reset", "value", counter.GetValue())
			return catena.OK(buildCounterResponse())
		},
	}

	// ==========================================================================
	// Assets (for GetAsset endpoint)
	// ==========================================================================
	assets := &sync.Map{}
	assetsList := &sync.Map{}

	// Load assets from embedded static directory
	loadAssetsFromEmbedded(staticFS, "static", assets, assetsList)

	// Build assets list for logging
	var assetNames []string
	assetsList.Range(func(key, value any) bool {
		assetNames = append(assetNames, key.(string))
		return true
	})

	// ==========================================================================
	// Server Setup
	// ==========================================================================
	slotList := []int{0}
	srv = rest.NewServer(slotList)

	// --------------------------------------------------------------------------
	// Register GetDevice handler
	// --------------------------------------------------------------------------
	srv.RegisterGetDeviceHandler(0, func(w http.ResponseWriter, r *http.Request) catena.StatusResult {
		parts := strings.Split(strings.Trim(r.URL.Path, "/"), "/")
		if len(parts) < 3 {
			return catena.BadRequest("invalid path")
		}
		slotStr := parts[2]
		slot, err := strconv.Atoi(slotStr)
		if err != nil {
			return catena.BadRequest("invalid slot")
		}

		logger.Info("GetDevice", "slot", slot)
		device, ok := devices[slot]
		if !ok {
			return catena.NotFound("device not found")
		}
		return catena.OK(device)
	})

	// --------------------------------------------------------------------------
	// Register GetValue handler (GetParam)
	// --------------------------------------------------------------------------
	srv.RegisterGetValueHandler(0, func(slot int, fqoid string) catena.StatusResult {
		logger.Info("GetParam", "slot", slot, "fqoid", fqoid)

		// Special case: return counter state
		if fqoid == "counter" {
			return catena.OK(buildCounterResponse())
		}

		// Special case: return all params as a struct
		if fqoid == "all" {
			allParams := make(map[string]any)
			params.Range(func(key, value any) bool {
				allParams[key.(string)] = value
				return true
			})
			allParams["counter"] = counter.GetValue()
			allParams["counter_running"] = counter.IsRunning()
			return catena.OK(allParams)
		}

		val, ok := params.Load(fqoid)
		if !ok {
			return catena.NotFound("parameter not found: " + fqoid)
		}

		return catena.OK(val)
	})

	// --------------------------------------------------------------------------
	// Register SetValue handler (SetParam)
	// --------------------------------------------------------------------------
	srv.RegisterSetValueHandler(0, func(value any, slot int, fqoid string) catena.StatusResult {
		logger.Info("SetParam", "slot", slot, "fqoid", fqoid, "value", value)

		// Check if parameter exists
		_, exists := params.Load(fqoid)
		if !exists {
			return catena.NotFound("parameter not found: " + fqoid)
		}

		params.Store(fqoid, value)
		logger.Info("Parameter updated", "fqoid", fqoid, "value", value)
		return catena.OK(value)
	})

	// --------------------------------------------------------------------------
	// Register ExecuteCommand handler
	// --------------------------------------------------------------------------
	srv.RegisterExecuteCommandHandler(0, func(w http.ResponseWriter, r *http.Request, slot int, commandFqoid string, payload any) catena.StatusResult {
		logger.Info("ExecuteCommand", "slot", slot, "command", commandFqoid)

		handler, ok := commands[commandFqoid]
		if !ok {
			logger.Warning("Command not found", "slot", slot, "command", commandFqoid)
			return catena.OK(map[string]any{
				"exception": map[string]any{
					"type":    "Invalid Command",
					"details": "Command not found: " + commandFqoid,
				},
			})
		}

		return handler(payload)
	})

	// --------------------------------------------------------------------------
	// Register GetAsset handler
	// --------------------------------------------------------------------------
	srv.RegisterGetAssetHandler(0, func(w http.ResponseWriter, r *http.Request, slot int, fqoid string) catena.StatusResult {
		logger.Info("GetAsset", "slot", slot, "fqoid", fqoid)

		val, ok := assets.Load(fqoid)
		if !ok {
			return catena.NotFound("asset not found: " + fqoid)
		}

		asset := val.(Asset)
		w.Header().Set("Content-Type", asset.ContentType)
		w.Header().Set("Content-Length", strconv.Itoa(len(asset.Data)))

		reader := bytes.NewReader(asset.Data)
		if _, err := io.Copy(w, reader); err != nil {
			return catena.InternalServerError("failed to stream asset")
		}

		return catena.StatusResult{Status: http.StatusOK}
	})

	// --------------------------------------------------------------------------
	// Register NotFound handler - serves index.html at root
	// --------------------------------------------------------------------------
	srv.RegisterNotFoundHandler(func(w http.ResponseWriter, r *http.Request) catena.StatusResult {
		if r.URL.Path == "/" {
			data, err := staticFS.ReadFile("static/index.htm")
			if err != nil {
				return catena.NotFound("index.html not found")
			}
			w.Header().Set("Content-Type", "text/html; charset=utf-8")
			w.Write(data)
			return catena.StatusResult{Status: http.StatusOK}
		}
		return catena.NotFound("endpoint not found: " + r.URL.Path)
	})

	// ==========================================================================
	// Start Server
	// ==========================================================================
	logger.Info("=======================================================")
	logger.Info("One of Everything REST Example")
	logger.Info("=======================================================")
	logger.Info("REST server starting", "port", port)
	logger.Info("")
	logger.Info("Web UI available at:")
	logger.Info(fmt.Sprintf("  http://localhost:%d/", port))
	logger.Info("")
	logger.Info("API endpoints:")
	logger.Info("  GET  /st2138-api/v1/0              - GetDevice")
	logger.Info("  GET  /st2138-api/v1/0/value/{oid}  - GetParam")
	logger.Info("  PUT  /st2138-api/v1/0/value/{oid}  - SetParam")
	logger.Info("  POST /st2138-api/v1/0/command/{cmd}- ExecuteCommand")
	logger.Info("  GET  /st2138-api/v1/0/asset/{oid}  - GetAsset")
	logger.Info("")
	logger.Info("Parameters: video/brightness, video/contrast, video/saturation,")
	logger.Info("            video/resolution, audio/volume, audio/muted, system/device_name")
	logger.Info("Commands:   start, stop, add10, reset")
	logger.Info("Assets:", "count", len(assetNames))
	logger.Info("=======================================================")

	srv.StartHTTPServer(port)

	// Wait for shutdown signal
	<-shutdownChan
	logger.Info("Server shutdown complete")
}

// loadAssetsFromEmbedded loads files from embedded FS into the asset store
func loadAssetsFromEmbedded(embedFS embed.FS, root string, assets *sync.Map, assetsList *sync.Map) {
	err := fs.WalkDir(embedFS, root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			return nil
		}
		// Skip index.html
		if filepath.Base(path) == "index.htm" {
			return nil
		}

		data, err := embedFS.ReadFile(path)
		if err != nil {
			return nil
		}

		ext := filepath.Ext(path)
		contentType := mime.TypeByExtension(ext)
		if contentType == "" {
			contentType = "application/octet-stream"
		}

		relPath, _ := filepath.Rel(root, path)
		assetID := strings.ReplaceAll(relPath, string(filepath.Separator), "/")

		assets.Store(assetID, Asset{
			ContentType: contentType,
			Data:        data,
			Filename:    filepath.Base(path),
		})
		assetsList.Store(assetID, true)
		logger.Info("Loaded asset", "id", assetID, "size", len(data))

		return nil
	})

	if err != nil {
		logger.Error("Error loading assets", "error", err)
	}
}

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}
